public/app/plugins/datasource/elasticsearch/utils.ts

```diff
@@ -106,3 +108,64 @@ export const unsupportedVersionMessage =
 export const isTimeSeriesQuery = (query: ElasticsearchQuery): boolean => {
   return query?.bucketAggs?.slice(-1)[0]?.type === 'date_histogram';
 };
+
+/*
+ * This regex matches 3 types of variable reference with an optional format specifier
+ * There are 6 capture groups that replace will return
+ * \$(\w+)                                    $var1
+ * \[\[(\w+?)(?::(\w+))?\]\]                  [[var2]] or [[var2:fmt2]]
+ * \${(\w+)(?:\.([^:^\}]+))?(?::([^\}]+))?}   ${var3} or ${var3.fieldPath} or ${var3:fmt3} (or ${var3.fieldPath:fmt3} but that is not a separate capture group)
+ */
+export const variableRegex = /\$(\w+)|\[\[(\w+?)(?::(\w+))?\]\]|\${(\w+)(?:\.([^:^\}]+))?(?::([^\}]+))?}/g;
+
+export const logLevelMap: Record<string, LogLevel> = {
+  emerg: LogLevel.emerg,
+  fatal: LogLevel.fatal,
+  alert: LogLevel.alert,
+  crit: LogLevel.crit,
+  critical: LogLevel.critical,
+  warn: LogLevel.warn,
+  warning: LogLevel.warning,
+  err: LogLevel.err,
+  eror: LogLevel.eror,
+  error: LogLevel.error,
+  info: LogLevel.info,
+  information: LogLevel.information,
+  informational: LogLevel.informational,
+  notice: LogLevel.notice,
+  dbug: LogLevel.dbug,
+  debug: LogLevel.debug,
+  trace: LogLevel.trace,
+  unknown: LogLevel.unknown,
+}
+
+export function dataFrameLogLevel(dataFrame: DataFrame): LogLevel {
+  const field = dataFrame.fields.find((f) => f.type === FieldType.number);
+  const level = field?.labels?.['level'] ?? '';
+  if (!level) {
+    return LogLevel.unknown;
+  }
+  return level in logLevelMap ? logLevelMap[level] : LogLevel.unknown;
+}
+
+export function flattenObject(target: Record<string, unknown> | unknown, prevKey = ''): Record<string, unknown> {
```

Comment:
Could we copy-paste here the logic from `flatten`. Cause this change is not equal, as it doesn't consider `maxDepth`. We are going to remove it with https://github.com/grafana/grafana/issues/65813 anyway, but as this PR shouldn't change the functionality, I would suggest to keep the original logic.

Reply:
Absolutely, will do that instead.

Reply:
Done.